pkg/ext/resources/types/types.go

```diff
@@ -0,0 +1,84 @@
+package types
+
+import (
+	"context"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/authentication/user"
+)
+
+type WatchEvent[T runtime.Object] struct {
+	Event  watch.EventType
+	Object T
+}
+
+// XXX: This is similar to k8s.io/apiserver/pkg/registry/rest.Standard, we could
+// aim to be somewhat compatible if we ever want to switch to using that. 🤷
+
+type Store[T runtime.Object, TList runtime.Object] interface {
+	Create(ctx context.Context, userInfo user.Info, obj T, opts *metav1.CreateOptions) (T, error)
+	Update(ctx context.Context, userInfo user.Info, obj T, opts *metav1.UpdateOptions) (T, error)
+	Get(ctx context.Context, userInfo user.Info, name string, opts *metav1.GetOptions) (T, error)
+	List(ctx context.Context, userInfo user.Info, opts *metav1.ListOptions) (TList, error)
+	// XXX: There's no way to safely close the channel outside of this call.
+	// We should probably return an interface instead like k8s.io/apimachinery/pkg/watch.Interface
+	Watch(ctx context.Context, userInfo user.Info, opts *metav1.ListOptions) (<-chan WatchEvent[T], error)
+	Delete(ctx context.Context, userInfo user.Info, name string, opts *metav1.DeleteOptions) error
+}
+
+type TableConvertor[TList runtime.Object] interface {
+	ConvertToTable(list TList, opts *metav1.TableOptions) *metav1.Table
+}
+
+type backingStore[T runtime.Object, TList runtime.Object] struct {
+	createFunc func(ctx context.Context, userInfo user.Info, obj T, opts *metav1.CreateOptions) (T, error)
+	updateFunc func(ctx context.Context, userInfo user.Info, obj T, opts *metav1.UpdateOptions) (T, error)
+	getFunc    func(ctx context.Context, userInfo user.Info, name string, opts *metav1.GetOptions) (T, error)
+	listFunc   func(ctx context.Context, userInfo user.Info, opts *metav1.ListOptions) (TList, error)
+	watchFunc  func(ctx context.Context, userInfo user.Info, opts *metav1.ListOptions) (<-chan WatchEvent[T], error)
+	deleteFunc func(ctx context.Context, userInfo user.Info, name string, opts *metav1.DeleteOptions) error
```

Comment:
The function types defined here are effectively repeated later in lines 69-74.
Would it make sense to create a set of named types and then use just these names here and 69-74 ?
Might be more readable.

Pity, I do not believe that such types could be used in the interface above (lines 21-28),
which is effectively the third copy of these function signatures.